### PR TITLE
WIP: Alert banner

### PR DESCRIPTION
**Pull request by pra7eek**
Closes - https://github.com/orgs/adobe/projects/19/views/18?filterQuery=alert&pane=issue&itemId=16308852

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-spectrum/alertbanner/src/AlertBanner.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import styles from '@adobe/spectrum-css-temp/components/alertbanner/vars.css';
+import {useProviderProps} from '@react-spectrum/provider';
+
+
+export interface SpectrumAlertBannerProps extends DOMProps, StyleProps {
+  text?: string | ReactNode,
+  variant?: 'neutral' | 'info' | 'negative',
+  actionLabel?: string,
+  onAction?: () => void,
+  isDismissable?: boolean
```
Comment (snowystinger) [inline packages/@react-spectrum/alertbanner/src/AlertBanner.tsx:29]: missing `onDismiss?: () => void`
which, if provided, would mean it's dismissable, so don't need the `isDismissable` prop, can remove it

File: packages/@react-spectrum/alertbanner/src/AlertBanner.tsx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import React, {ReactNode} from 'react';
+import styles from '@adobe/spectrum-css-temp/components/alertbanner/vars.css';
+import {useProviderProps} from '@react-spectrum/provider';
+
+
+export interface SpectrumAlertBannerProps extends DOMProps, StyleProps {
+  text?: string | ReactNode,
+  variant?: 'neutral' | 'info' | 'negative',
+  actionLabel?: string,
+  onAction?: () => void,
```
Comment (snowystinger) [inline packages/@react-spectrum/alertbanner/src/AlertBanner.tsx:28]: possibly need a `shouldCloseOnAction` much like Toast has

File: packages/@react-spectrum/alertbanner/src/AlertBanner.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      role="alert">
+      {Icon &&
+        <Icon
+          aria-label={iconLabel}
+          UNSAFE_className={classNames(styles, 'spectrum-AlertBanner-typeIcon')} />
+      }
+      <div className={classNames(styles, 'spectrum-AlertBanner-body')}>
+        <div className={classNames(styles, 'spectrum-AlertBanner-content')} >{text || ''}</div>
+        {actionLabel &&
+        <Button
```
Comment (snowystinger) [inline packages/@react-spectrum/alertbanner/src/AlertBanner.tsx:73]: why is the action button inside the body?
everything should be a child of the AlertBanner and just laid out in a non-wrapping flex with
`flex: 0 0 auto` on everything except the content text which is `flex: 1 1 auto; min-width: 0;`

File: packages/@adobe/spectrum-css-temp/components/alertbanner/index.css [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  }
+}
+
+.spectrum-AlertBanner-body {
+  flex: 1 1 auto;
+  align-self: center;
+
+  /* https://spectrum.adobe.com/page/alert-banner/#Text-overflow */
+  display: flex;
+  flex-wrap: wrap;
```
Comment (snowystinger) [inline packages/@adobe/spectrum-css-temp/components/alertbanner/index.css:68]: text overflow is different, i don't think we want wrap here
this should all go away though with my other comment about moving the action button up so everything is a sibiling

File: packages/@react-spectrum/alertbanner/src/AlertBanner.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  onAction?: () => void,
+  isDismissable?: boolean
+}
+
+const ICONS = {
+  info: InfoMedium,
+  negative: AlertMedium
+};
+
+function AlertBanner(props: SpectrumAlertBannerProps, ref: DOMRef<HTMLDivElement>) {
```
Comment (snowystinger) [inline packages/@react-spectrum/alertbanner/src/AlertBanner.tsx:37]: question to team, is this a landmark?

File: packages/@react-spectrum/alertbanner/test/AlertBanner.test.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import {AlertBanner} from '../';
+import React from 'react';
+import {render} from '@react-spectrum/test-utils';
+
+describe('AlertBanner', function () {
+  it.each`
+    Name | Component      | props
+    ${'AlertBanner'} | ${AlertBanner} | ${{}}
+  `('$Name handles defaults', function ({Component, props}) {
+    let tree = render(<Component {...props} />);
```
Comment (snowystinger) [inline packages/@react-spectrum/alertbanner/test/AlertBanner.test.js:22]: would be good to see some tests
for instance, onDismiss and onAction
ref

**Review by snowystinger — COMMENTED** [review]
questions to team, what is a non-dismissable alert banner? how does it get removed?
should we have an banneralert container like we do toasts? queue?

File: packages/@react-spectrum/alertbanner/src/AlertBanner.tsx (hunk 5 above)
Comment (reidbarber) [reply packages/@react-spectrum/alertbanner/src/AlertBanner.tsx:37]: There is a banner landmark: https://www.w3.org/WAI/ARIA/apg/patterns/landmarks/examples/banner.html

> A banner landmark identifies site-oriented content at the beginning of each page within a website. Site-oriented content typically includes things such as the logo or identity of the site sponsor, and site-specific search tool. A banner usually appears at the top of the page and typically spans the full width.

Maybe a question for the accessibility team on whether this qualifies.

**Comment by snowystinger** [comment]
> https://spectrum.adobe.com/page/alert-banner/#Consider-including-a-close-button

This also says action is optional. So we still need a way for this to close. I propose that dismissable is not actually optional and that the design is wrong. For the same reason as Toast, I don't think people should be removing this based on a timer so that users have all the time they need to get to the banner and interact with it and then dismiss it themselves.

**Comment by reidbarber** [comment]
True, looks like timers should be avoided, but it should be able to be dismissed programmatically if an issue is resolved:

> Never allow alert banners to time out. Since these are system-level messages, they should not dismiss on their own unless there is a change in the system that resolves an issue (e.g., regaining internet connectivity after losing it).

https://spectrum.adobe.com/page/alert-banner/#Dont-let-alert-banners-time-out

**Comment by snowystinger** [comment]
Right, I think this supports my suggestion that they are *always* dismissable. The system may not resolve and the user may want their full screen back. It's the same behavior as our Toasts, they are always dismissable but they can be programmatically dismissed as well.

We could probably reuse/copy the ToastContainer/Queue for this and follow the same work for a landmark region.

@devongovett @dannify please have a think as well about the direction we want to go with this.

**Review by snowystinger — COMMENTED** [review]
Team discussion results:
We will not implement a toast-like container for this. Component will just visually look like an AlertBanner, clicking the dismiss button just calls a onDismiss prop, user must remove it themselves. They also must manage positioning and placement of the component. So this PR is already very close to that, one of us will review again with all of this in mind. I'll resolve all my open ones in a little bit.

File: packages/@react-spectrum/alertbanner/src/AlertBanner.tsx (hunk 2 above)
Comment (snowystinger) [reply packages/@react-spectrum/alertbanner/src/AlertBanner.tsx:28]: Update: no longer needed as the onAction and onDismiss would just be things called and the user would need to handle the dismissal in both cases depending on if it should dismiss or not

File: packages/@react-spectrum/alertbanner/src/AlertBanner.tsx (hunk 5 above)
Comment (snowystinger) [reply packages/@react-spectrum/alertbanner/src/AlertBanner.tsx:37]: resolving this one, we're not worrying at this at the moment

**Comment by snowystinger** [comment]
GET_BUILD

**Comment by snowystinger** [comment]
Closing for same reason as https://github.com/adobe/react-spectrum/pull/6780#issuecomment-2270344183
We can reopen when we have that information
Thanks for your understanding

**Pull request closed without merging**